javalib/src/main/scala/java/util/regex/PatternCompiler.scala

```diff
@@ -0,0 +1,1645 @@
+      parseError("Hexadecimal codepoint is too big")
+
+    cp
+  }
+
+  /** Parses and returns a translated version of a pre-defined character class. */
+  private def parsePredefinedCharacterClass(dispatchChar: Char): CompiledCharClass = {
+    import CompiledCharClass._
+
+    pIndex += 1
+
+    val positive = (dispatchChar: @switch) match {
+      case 'd' | 'D' =>
+        if (unicodeCharacterClass) UnicodeDigit
+        else ASCIIDigit
+      case 'h' | 'H' =>
+        UniversalHorizontalWhiteSpace
+      case 's' | 'S' =>
+        if (unicodeCharacterClass) UnicodeWhitespace
+        else ASCIIWhiteSpace
+      case 'v' | 'V' =>
+        UniversalVerticalWhiteSpace
+      case 'w' | 'W' =>
+        wordCharClass
+      case 'p' | 'P' =>
+        parsePCharacterClass()
+    }
+
+    if (dispatchChar >= 'a')
+      positive
+    else
+      positive.negated
+  }
+
+  /** Parses and returns a translated version of a `\p` character class. */
+  private def parsePCharacterClass(): CompiledCharClass = {
+    import CompiledCharClass._
+
+    val pattern = this.pattern // local copy
+    val len = pattern.length()
+
+    val start = pIndex
+    val property = if (start == len) {
+      "?" // mimics the behavior of the JVM
+    } else if (pattern.charAt(start) == '{') {
+      val innerStart = start + 1
+      val innerEnd = pattern.indexOf("}", innerStart)
+      if (innerEnd < 0)
+        parseError("Unclosed character family")
+      pIndex = innerEnd
+      pattern.substring(innerStart, innerEnd)
+    } else {
+      pattern.substring(start, start + 1)
+    }
+
+    val result = if (!unicodeCharacterClass && asciiPOSIXCharacterClasses.contains(property)) {
+      val property2 =
+        if (asciiCaseInsensitive && (property == "Lower" || property == "Upper")) "Alpha"
+        else property
+      asciiPOSIXCharacterClasses(property2)
+    } else {
+      // For anything else, we need built-in support for \p
+      requireES2018Features("Unicode character family")
+
+      predefinedCharacterClasses.getOrElse(property, {
+        val scriptPrefixLen = if (property.startsWith("Is")) {
+          2
+        } else if (property.startsWith("sc=")) {
+          3
+        } else if (property.startsWith("script=")) {
+          7
+        } else if (property.startsWith("In") || property.startsWith("blk=") || property.startsWith("block=")) {
+          parseError("Blocks are not supported in \\p Unicode character families")
+        } else {
+          // Error
+          parseError(s"Unknown Unicode character class '$property'")
+        }
+        posP("sc=" + canonicalizeScriptName(property.substring(scriptPrefixLen)))
+      })
+    }
+
+    pIndex += 1
+
+    result
+  }
+
+  private def compileCharacterClass(): String = {
+    // scalastyle:off return
+
+    val impossiblePattern = "[^\\d\\D]"
+
+    // State that crosses functions
+    val state = new CharacterClassState(asciiCaseInsensitive)
+    import state._
```

Comment:
I just saw that... I don't see the value of this. Either:
- Put all the state locally.
- Create a character class builder that protects its internal state.